### ISSUE TITLE
감정분류

### ISSUE THREAD

**Issue by taeo-daddy** [taeo-daddy/aice_pro#1]
### 텍스트 
1 .AI-HUB 감정말뭉치 데이터 읽어오고
2.데이터 전처리 : 한글, 공백외의 영어, 숫자, 특수문자등 제거
3.TF-IDF 토큰나이져 활용하여 토큰화하고 문장을 숫자로 나열
4.머신러닝 RandomForest 모델을 활용하여 감성분류 수행

**import numpy as np
import pandas as pd**

AI-HUB 감성 대화 말뭉치 활용하여 만든 데이터 읽어오기
**final_data = pd.read_csv('https://github.com/ohgzone/file1/raw/main/aihub_coupus.csv' )**

데이터 확인하기
**final_data.head()**

총 51,630건
**final_data.info()**

'문장' 컬럼의 내용중에 영문, 특수문자 있는지 확인 : 영문과 특수문
**final_data[final_data['문장'].str.contains('[^가-힣 ]')].values**

'문장' 컬럼의 내용에서 숫자, 영문자, 특수문자등의 글자는 삭제처리
**final_data['문장'] = final_data['문장'].str.replace('[^가-힣 ]','', regex=True)**

'문장' 컬럼의 내용에서 영문, 특수문자 없음 확인
**final_data[final_data['문장'].str.contains('[^가-힣 ]')].sum()**

숫자, 영문자, 특수문자 등 제거후 데이터 확인하기.
**final_data.head()**

final_data 어떤 컬럼과 내용으로 되어 있는지 다시 확인
**final_data.tail()**

'문장' 컬럼의 내용을 양끝의 빈공간 삭제
**final_data['문장'] = final_data['문장'].str.strip()**

Null 있는지 확인 : 없음
**final_data.isnull().sum()**

중복 데이터 있는지 확인 : 56건 중복 존재 확인
**final_data['문장'].duplicated().sum()**

중복 데이터 제거
**final_data.drop_duplicates(subset=['문장'], inplace=True)**

label '감정' 분포 확인 : 총 6개이며, 고루게 분포 확인. 단 기쁨이 약간 부족해 보임
**final_data['감정'].value_counts()**

라벨와 클래스을 매핑 작업
**from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
final_data['감정'] = encoder.fit_transform(final_data['감정'])
encoder.classes_**

X, Y 분리
**features = final_data['문장'].values
labels = final_data['감정'].values**

**print('이벤트 문자열 최대 길이 :{}'.format(max(len(l) for l in features)))
print('이벤트 문자열 평균 길이 :{}'.format(sum(map(len, features))/len(features)))**

히스토그램을 보면 30~40 부근에 많이 몰려 있음 알수 있다.
**plt.hist([len(s) for s in features], bins=50)
plt.xlabel('length of samples')
plt.ylabel('number of samples')
plt.show()**

train set와 test set 분리
**from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(features, labels , test_size=0.2, stratify=labels, random_state=41)**

**x_train.shape, x_test.shape, y_train.shape, y_test.shape**

말뭉치를 TF-IDF로 변환하기
**from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer()
x_train_v = tfidf.fit_transform(x_train)
x_test_v = tfidf.transform(x_test)**

각 라인의 각 단어에 대한 TF-IDF 값 표현
**print(x_train_v)**

머신러닝 모델링
학습하는데 Colab에서 4분 소요
**from sklearn.ensemble import RandomForestClassifier**

예측해보기
출력 결과 해석 : (0, 44327)	0.241660101642553
0 : 첫라인, 44327 : 단어에 맵핑된 번호, 0.241660101642553 : tf-idf 계산 값

RandomForest 모델로 예측하기
**predict = rfc.predict(x_test_v[:1])
predict, encoder.inverse_transform(predict)**

**print(f'검증데이터셋의 첫번째 TF-IDF : {x_test_v[0]}')
rfc = RandomForestClassifier() # 모델 정의
rfc.fit(x_train_v, y_train) # 모델 학습
rfc.score(x_test_v, y_test)**

===========================================================
### 텍스트
1.AI-HUB 감정말뭉치 데이터 읽어오고
2.데이터 전처리 : 한글, 공백외의 영어, 숫자, 특수문자등 제거
3.keras 토큰나이져 활용하여 토큰화하고 문장을 숫자로 나열(texts_tp_sequences)
4.일정 크기의 문장으로 맞추기 위해 패딩수행(pad_sequences)
5.embedding 레이어를 이용해서 단어를 n차원 밀집벡터를 만들고 train시에 학습되도록 했숩니다.
6.LSTM 언어 모델을 활용하여 감성분류 수행

전처리까지 동일함 / label 숫자로 인코딩부터 시작

감정 리스트 만듬
**list1 = final_data['감정'].value_counts().index.values
list1**

라벨와 클래스을 매핑 작업
**label2class = {}
class2label = {}
for cl, la in enumerate(list1):
  label2class[la] = cl
  class2label[cl] = la**

**print(label2class)
print(class2label)**

'감정' 라벨링 수행
**final_data['label'] = final_data['감정'].map(label2class)**

X, Y 분리
**features = final_data['문장'].values
labels = final_data['label'].values**

**print('이벤트 문자열 최대 길이 :{}'.format(max(len(l) for l in features)))
print('이벤트 문자열 평균 길이 :{}'.format(sum(map(len, features))/len(features)))**

from sklearn.model_selection import train_test_split

**x_train, x_test, y_train, y_test = train_test_split(features, labels , test_size=0.2, stratify=labels, random_state=41)
x_train.shape, x_test.shape, y_train.shape, y_test.shape**

샘플확인 , 라벨 확인
{0: '불안', 1: '분노', 2: '상처', 3: '슬픔', 4: '당황', 5: '기쁨'}
**x_train[:2], y_train[:2]**

Tokenizing
**import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences**

Tokenizer 구현 : 단어 사전 만들기(fit_on_texts)
**tokenizer = Tokenizer()
tokenizer.fit_on_texts(x_train)**

단어에 대한 숫자 매핑
**print(tokenizer.word_index)**

반대로 숫자로 단어 매핑
**print(tokenizer.index_word)**

단어별 빈도수 확인
**print(tokenizer.word_counts)**

총 단어 갯수 : 47,646
**max_words = len(tokenizer.index_word)
print(max_words)**

texts_to_sequences : 문장을 숫자로 나열
문장을 숫자로 나열
**x_train_seq = tokenizer.texts_to_sequences(x_train)
x_test_seq = tokenizer.texts_to_sequences(x_test)**

문장을 숫자로 변경후 갯수 확인
x_train.shape, x_test.shape, y_train.shape, y_test.shape : ((41259,), (10315,), (41259,), (10315,))
**print(len(x_train_seq), len(x_test_seq))**

11. Padding Sequence
문장의 최대 길이 파악 : 제일 긴 문장 seq 길이는 38개로 구성됨.
**max(len(line) for line in x_train_seq)**

모든 문장을 최대 문장 Seq 길이 38에 맞춘다.
**x_train_pad = pad_sequences(x_train_seq, maxlen=38)
x_test_pad = pad_sequences(x_test_seq, maxlen=38)**

문장 Seq 내용을 보니 잘 패딩되어 있음 확인
**x_train_pad[:1]**

문장 Seq 패딩의 shape 확인
**x_train_pad.shape, x_test_pad.shape**

12. LSTM 모델링
**from tensorflow.keras.layers import Dense, Flatten, Conv1D, MaxPool2D
from tensorflow.keras.layers import Embedding, Bidirectional, LSTM, SimpleRNN, GRU
from tensorflow.keras.models import Sequential
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint**

하이퍼 파라미터
**max_words = 47646 + 1    # 총 단어 갯수 + padding 0 번호
max_len = 38             # 최대 문장 길이
embedding_dim = 32      # embedding 차원**

모델 선언
**model =  Sequential()**

단어를 의미있는 32 차원으로 Vector 변경(Embedding)
**model.add(Embedding(max_words, embedding_dim, input_length=max_len))**

LSTM 모델
**model.add(LSTM(16, return_sequences=True))
model.add(LSTM(16, return_sequences=True))
model.add(Flatten())
model.add(Dense(128, activation='swish'))
model.add(Dense(32, activation='swish'))
model.add(Dense(6, activation='softmax'))**

모델 compile
**model.compile(loss = 'sparse_categorical_crossentropy',
              optimizer = 'adam',
              metrics = ['accuracy'])
model.summary()**

조기종료 콜백함수 정의(EarlyStopping)
**es =  EarlyStopping(monitor='val_loss', patience=10, verbose=1)**

체크포인트 저장(ModelCheckpoint)
**checkpoint_path = 'tmp_checkpoint.keras'
cp =  ModelCheckpoint(checkpoint_path, monitor='val_loss', verbose=1, save_best_only=True)**

모델 학습(fit)
**history = model.fit(x_train_pad, y_train, epochs=50, batch_size=512,
                      validation_split=0.2, verbose =1, callbacks=[es, cp])**

**epochs = range(1, len(history.history['accuracy']) + 1)
plt.plot(epochs, history.history['accuracy'])
plt.plot(epochs, history.history['val_accuracy'])
plt.title('model accuracy')
plt.ylabel('loss')
plt.xlabel('epoch')
plt.legend(['train', 'valid'], )
plt.show()**

**model.evaluate(x_test_pad, y_test)**

13. 예측해보기

**print(f'문자열 : {x_test[0]}')
print(f'Sequence : {x_test_pad[0]}')**

모델 예측하기(predict)
**predict = model. [칸 채우기] (x_test_pad[:1])**

**print(f'True : {class2label[y_test[0]]}')
print(f'Predict : {class2label[np.argmax(predict)]}')**
